server/pkcs11.go

```diff
@@ -0,0 +1,172 @@
+package server
+
+import (
+	"crypto"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cbroglie/crypto11"
+)
+
+// The PKCS#11 URI is a scheme for identifying PKCS#11 objects stored on PKCS#11
+// tokens and how to access them.
+// For more information read: https://tools.ietf.org/html/rfc7512#section-2.3
+type PKCS11URI struct { // pkcs11:path-attr[?query-attr]
+	// path-attr: (; delimited)
+	Token  string //        token <- CK_TOKEN_INFO
+	Manuf  string // manufacturer <- CK_TOKEN_INFO
+	Serial string //       serial <- CK_TOKEN_INTO
+	Model  string //        model <- CK_TOKEN_INFO
+
+	LibManuf string // library-manufacturer <- CK_INFO
+	LibDesc  string //  library-description <- CK_INFO
+	LibVer   string //      library-version <- CK_INFO
+
+	Object []byte // object <- CKA_LABEL
+	Type   string //   type <- CKA_CLASS
+	ID     []byte //     id <- CKA_ID
+
+	SlotManuf string // slot-manufacturer <- CK_SLOT_INFO
+	SlotDesc  string //  slot-description <- CK_SLOT_INFO
+	SlotID    uint   //           slot-id <- CK_SLOT_ID
+
+	// query-attr: (& delimited)
+	PinSource string // pin-source
+	PinValue  string //  pin-value
+
+	ModuleName string // module-name
+	ModulePath string // module-path
+
+	// Vendor specific query attributes:
+	MaxSessions int // max-sessions
+}
+
+// This function parses PKCS#11 URIs defined in RFC 7512 into PKCS11URI objects.
+// An error is returned if the input string does not appear to follow the rules
+// or if there are unrecognized components.
+func RFC7512Parser(uri string) (*PKCS11URI, error) {
```

Comment:
The function name is bad, and the function docs should start with "FunctionName parses...", and the function docs are currently a restatement of the function signature.